src/Microsoft.Extensions.Hosting.Abstractions/BackgroundService.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Hosting
+{
+    /// <summary>
+    /// Base class for implementing a long running <see cref="IHostedService"/>.
+    /// </summary>
+    public abstract class BackgroundService : IHostedService, IDisposable
+    {
+        private static Action<object> _cancelToken = CancelToken;
+
+        private Task _executingTask;
+        private CancellationTokenSource _executionCts;
+        private readonly CancellationTokenSource _shutdownCts = new CancellationTokenSource();
+
+        /// <summary>
+        /// This method is called when the <see cref="IHostedService"/> starts. The implementation should return a task that represents
+        /// the lifetime of the long running operation(s) being performed.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> that fires when the <see cref="IHostedService"/> is stopped.</param>
+        /// <returns>A <see cref="Task"/> that represents the long running operations.</returns>
+        protected abstract Task ExecuteAsync(CancellationToken cancellationToken);
+
+        /// <summary>
+        /// A <see cref="CancellationToken"/> that repesents the ungraceful shutdown.
+        /// </summary>
+        public CancellationToken ShutdownToken => _shutdownCts.Token;
```

Comment:
AbortToken?

Reply:
I'm not completely clear how this can be used. When would this fire? If it is an ungraceful shutdown how can anyone have time to react to it? Otherwise it would be graceful. Would it not?

Reply:
Consider the WebHost 5 second shutdown timeout we have today. StopAsync is graceful until that timeout expires, then it becomes ungracefull and kestrel starts killing sockets.

Reply:
The process still waits for Kestrel to finish before shutting down, it just asks Kestrel to hurry it up.

Reply:
So it is more like a Stopping/Stop scenario where ShutdownToken is stop and the other graceful one is, we are stopping start cleaning up?

Reply:
Yes.

Reply:
It's really not going to be used much. This for the 0.000001% case that you want to handle the shutdown timeout after being signalled for stopping in the first place. I'm leaving it as it is.